modal/_container_entrypoint.py

```diff
@@ -444,27 +487,39 @@ async def call_function_async(
                 obj.__enter__()
 
     try:
-        async for input_id, args, kwargs in function_io_manager.run_inputs_outputs.aio():
+
+        async def run_input(input_id, args, kwargs):
             output_index = SequenceNumber(0)  # mutable number we can increase from the generator loop
-            async with function_io_manager.handle_input_exception.aio(input_id, output_index):
+            started_at = time.time()
+            async with function_io_manager.handle_input_exception.aio(input_id, started_at, output_index):
                 res = fun(*args, **kwargs)
 
                 # TODO(erikbern): any exception below shouldn't be considered a user exception
                 if is_generator:
                     if not inspect.isasyncgen(res):
                         raise InvalidError(f"Async generator function returned value of type {type(res)}")
                     async for value in res:
-                        await function_io_manager.enqueue_generator_value.aio(input_id, output_index.value, value)
+                        await function_io_manager.enqueue_generator_value.aio(
+                            input_id, started_at, output_index.value, value
+                        )
                         output_index.increase()
-                    await function_io_manager.enqueue_generator_eof.aio(input_id, output_index.value)
+                    await function_io_manager.enqueue_generator_eof.aio(input_id, started_at, output_index.value)
                 else:
                     if not inspect.iscoroutine(res) or inspect.isgenerator(res) or inspect.isasyncgen(res):
                         raise InvalidError(
                             f"Async (non-generator) function returned value of type {type(res)}"
                             " You might need to use @stub.function(..., is_generator=True)."
                         )
                     value = await res
-                    await function_io_manager.enqueue_output.aio(input_id, output_index.value, value)
+                    await function_io_manager.enqueue_output.aio(input_id, started_at, output_index.value, value)
+
+        if input_concurrency > 1:
+            async with TaskContext() as execution_context:
+                async for input_id, args, kwargs in function_io_manager.run_inputs_outputs.aio(input_concurrency):
+                    execution_context.create_task(run_input(input_id, args, kwargs))
```

Comment:
actually why do you need to create tasks here? can't you just run `asyncio.gather` on a bunch of coroutines?